src/pybind/mgr/cephadm/services/cephadmservice.py

```diff
@@ -636,6 +643,52 @@ def post_remove(self, daemon: DaemonDescription, is_failed_deploy: bool) -> None
         # super().post_remove(daemon)
         pass
 
+    def generate_config(self, daemon_spec: CephadmDaemonDeploySpec) -> Tuple[Dict[str, Any], List[str]]:
+        daemon_spec.final_config, daemon_spec.deps = super().generate_config(daemon_spec)
+
+        mon_spec = cast(MONSpec, self.mgr.spec_store[daemon_spec.service_name].spec)
+        if mon_spec.crush_locations:
+            if daemon_spec.host in mon_spec.crush_locations:
+                # the --crush-location flag only supports a single bucker=loc pair so
+                # others will have to be handled later. The idea is to set the flag
+                # for the first bucket=loc pair in the list in order to facilitate
+                # replacing a tiebreaker mon (https://docs.ceph.com/en/quincy/rados/operations/stretch-mode/#other-commands)
+                c_loc = mon_spec.crush_locations[daemon_spec.host][0]
+                daemon_spec.final_config['crush_location'] = c_loc
+
+        return daemon_spec.final_config, daemon_spec.deps
+
+    def set_crush_locations(self, daemon_descrs: List[DaemonDescription], spec: ServiceSpec) -> None:
+        if not daemon_descrs:
+            return
+        assert self.TYPE == spec.service_type
+        mon_spec = cast(MONSpec, spec)
+
+        if mon_spec.crush_locations:
+            quorum_status = self._get_quorum_status()
+            mons_in_monmap = [m['name'] for m in quorum_status['monmap']['mons']]
+            for dd in daemon_descrs:
+                assert dd.daemon_id is not None
+                assert dd.hostname is not None
+                if dd.hostname not in mon_spec.crush_locations:
+                    continue
+                if dd.daemon_id not in mons_in_monmap:
+                    continue
+                # expected format for crush_locations from the quorum status is
+                # {bucket1=loc1,bucket2=loc2} etc. for the number of bucket=loc pairs
+                current_crush_locs = [m['crush_location'] for m in quorum_status['monmap']['mons'] if m['name'] == dd.daemon_id][0]
```

Comment:
Does `[m['crush_location'] for m in quorum_status['monmap']['mons'] if m['name'] == dd.daemon_id]` always return something? otherwise we could try to access an empty list.

Reply:
I don't know for sure. It seemed to when I was testing. Anyway, I've added a try except around this just in case.